x/poolmanager/router_test.go

```diff
@@ -98,6 +98,147 @@ func (suite *KeeperTestSuite) TestGetPoolModule() {
 	}
 }
 
+func (suite *KeeperTestSuite) TestRouteGetPoolDenoms() {
+	tests := map[string]struct {
+		poolId            uint64
+		preCreatePoolType types.PoolType
+		routesOverwrite   map[types.PoolType]types.SwapI
+
+		expectedDenoms []string
+		expectError    error
+	}{
+		"valid balancer pool": {
+			preCreatePoolType: types.Balancer,
+			poolId:            1,
+			expectedDenoms:    []string{"bar", "baz", "foo", "uosmo"},
+		},
+		"valid stableswap pool": {
+			preCreatePoolType: types.Stableswap,
+			poolId:            1,
+			expectedDenoms:    []string{"bar", "baz", "foo"},
+		},
+		"valid concentrated liquidity pool": {
+			preCreatePoolType: types.Concentrated,
+			poolId:            1,
+			expectedDenoms:    []string{"eth", "usdc"},
+		},
+
+		"non-existent pool": {
+			preCreatePoolType: types.Balancer,
+			poolId:            2,
+
+			expectError: types.FailedToFindRouteError{PoolId: 2},
+		},
+		"undefined route": {
+			preCreatePoolType: types.Balancer,
+			poolId:            1,
+			routesOverwrite: map[types.PoolType]types.SwapI{
+				types.Stableswap: &gamm.Keeper{}, // undefined for balancer.
+			},
+
+			expectError: types.UndefinedRouteError{PoolId: 1, PoolType: types.Balancer},
+		},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		suite.Run(name, func() {
+			suite.SetupTest()
+			poolmanagerKeeper := suite.App.PoolManagerKeeper
+
+			suite.createPoolFromType(tc.preCreatePoolType)
+
+			if len(tc.routesOverwrite) > 0 {
+				poolmanagerKeeper.SetPoolRoutesUnsafe(tc.routesOverwrite)
+			}
+
+			denoms, err := poolmanagerKeeper.RouteGetPoolDenoms(suite.Ctx, tc.poolId)
+			if tc.expectError != nil {
+				suite.Require().Error(err)
+				suite.Require().ErrorIs(err, tc.expectError)
+				return
+			}
+			suite.Require().NoError(err)
+			suite.Require().Equal(tc.expectedDenoms, denoms)
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestRouteCalculateSpotPrice() {
+	tests := map[string]struct {
+		poolId            uint64
+		preCreatePoolType types.PoolType
+		quoteAssetDenom   string
+		baseAssetDenom    string
+
+		routesOverwrite   map[types.PoolType]types.SwapI
+		expectedSpotPrice sdk.Dec
+
+		expectError error
+	}{
+		"valid balancer pool": {
+			preCreatePoolType: types.Balancer,
+			poolId:            1,
+			quoteAssetDenom:   "bar",
+			baseAssetDenom:    "baz",
+			expectedSpotPrice: sdk.MustNewDecFromStr("1.5"),
+		},
+		"valid stableswap pool": {
+			preCreatePoolType: types.Stableswap,
+			poolId:            1,
+			quoteAssetDenom:   "bar",
+			baseAssetDenom:    "baz",
+			expectedSpotPrice: sdk.MustNewDecFromStr("0.99999998"),
+		},
+		"valid concentrated liquidity pool": {
+			preCreatePoolType: types.Concentrated,
+			poolId:            1,
+			quoteAssetDenom:   "usdc",
+			baseAssetDenom:    "eth",
+			expectedSpotPrice: sdk.OneDec(),
+		},
```

Comment:
Ahh I see why you added that spot price logic earlier, because when we initialize a CL pool there are no assets and it defaults to a sqrtPrice of zero. 

The logic you added though still needs to be removed, and I suggest for this test we just add a logic gate to initialize a position if its a concentrated pool types so we can get a valid spot price

Reply:
@czarcas7ic  When a pool is created and a position has been created with equal amounts of each asset, the current tick (sqrt price) would stay the same (=zero)

Reply:
This should not be the case. My guess is that you potentially did not re-fetch the pool when re-querying the spot price, so you got the same spot price as when the pool was created (I know because I have made this mistake a couple of times now haha)

Reply:
Edited the logic so that on zero spot price it would return zero as is (could not completely eliminate the logic gate of checking if spotPrice is not zero since this would lead to zero division errors)